builtin/providers/aws/resource_aws_instance.go

```diff
@@ -261,38 +261,87 @@ func resourceAwsInstanceCreate(d *schema.ResourceData, meta interface{}) error {
 		userData = v.(string)
 	}
 
+	placement := &ec2.Placement{
+		AvailabilityZone: aws.String(d.Get("availability_zone").(string)),
+		Tenancy:          aws.String(d.Get("tenancy").(string)),
+	}
+
+	iam := &ec2.IAMInstanceProfileSpecification{
+		Name: aws.String(d.Get("iam_instance_profile").(string)),
+	}
+
+	// Build the creation struct
+	runOpts := &ec2.RunInstancesRequest{
+		ImageID:            aws.String(d.Get("ami").(string)),
+		Placement:          placement,
+		InstanceType:       aws.String(d.Get("instance_type").(string)),
+		MaxCount:           aws.Integer(1),
+		MinCount:           aws.Integer(1),
+		UserData:           aws.String(userData),
+		EBSOptimized:       aws.Boolean(d.Get("ebs_optimized").(bool)),
+		IAMInstanceProfile: iam,
+	}
+
 	associatePublicIPAddress := false
 	if v := d.Get("associate_public_ip_address"); v != nil {
 		associatePublicIPAddress = v.(bool)
 	}
 
-	// Build the creation struct
-	runOpts := &ec2.RunInstances{
-		ImageId:                  d.Get("ami").(string),
-		AvailZone:                d.Get("availability_zone").(string),
-		InstanceType:             d.Get("instance_type").(string),
-		KeyName:                  d.Get("key_name").(string),
-		SubnetId:                 d.Get("subnet_id").(string),
-		PrivateIPAddress:         d.Get("private_ip").(string),
-		AssociatePublicIpAddress: associatePublicIPAddress,
-		UserData:                 []byte(userData),
-		EbsOptimized:             d.Get("ebs_optimized").(bool),
-		IamInstanceProfile:       d.Get("iam_instance_profile").(string),
-		Tenancy:                  d.Get("tenancy").(string),
+	// check for non-default Subnet
+	subnet := false
+	var subnetID string
+	if v, ok := d.GetOk("subnet_id"); ok {
+		subnet = true
+		subnetID = v.(string)
+	}
+
+	if subnet && associatePublicIPAddress {
+		// If we have a non-default VPC / Subnet specified, we can flag
+		// AssociatePublicIpAddress to get a Public IP assigned. By default these are not provided.
+		// You cannot specify both SubnetId and the NetworkInterface.0.* parameters though, otherwise
+		// you get: Network interfaces and an instance-level subnet ID may not be specified on the same request
+		// You also need to attach Security Groups to the NetworkInterface instead of the instance,
+		// to avoid: Network interfaces and an instance-level security groups may not be specified on
+		// the same request
```

Comment:
I coped this from `goamz`, which handled this logic for us internally, but now we need to be responsible for the magic so it seemed reasonable to copy the rationale